xpdan/data_reduction_core.py

```diff
@@ -0,0 +1,576 @@
+                        dark_hdr, fill=True))['data'][image_data_key]
+            elif name == 'descriptor':
+                pass
+            elif name == 'event':
+                f_name = _feature_gen(doc)
+                img = doc['data'][image_data_key]
+                if dark_sub_bool:
+                    img -= dark_img
+                path = os.path.join(save_dir, f_name)
+                if not dryrun:
+                    tif.imsave(path, img)
+                    if os.path.isfile(path):
+                        print('image "%s" has been saved at "%s"' %
+                              (f_name, path))
+                    else:
+                        print(
+                            'Sorry, something went wrong with your tif saving')
+                        return
+                else:
+                    print('dryrun: image "%s" has been saved at "%s"' %
+                          (f_name, path))
+            elif name == 'stop':
+                # save run_start
+                stem, ext = os.path.splitext(path)
+                config_name = stem + '.yml'
+                with open(config_name, 'w') as f:
+                    yaml.dump(header.start, f)  # save all md in start
+
+    print(" *** {} *** ".format('Saving process finished'))
+
+
+def save_last_tiff(**kwargs):
+    """Save images from latest data set as tiff format files.
+
+    Parameters
+    ----------
+    db: databroker.broker.Broker instance
+        The databroker holding the data
+    save_dir: str
+        The folder in which to save the data
+    dark_sub_bool : bool, optional
+        Default is True, which allows dark/background subtraction to
+        be done before saving each image. If header doesn't contain
+        necessary information to perform dark subtraction, uncorrected
+        image will be saved.
+    dryrun : bool, optional
+        if set to True, file won't be saved. default is False
+    image_data_key: str
+        The key for event['data'] which gives back the image
+    """
+
+    save_tiff(kwargs['db'][-1], **kwargs)
+
+
+def sum_images(event_stream, idxs_list=None):
+    """Sum images in a header
+
+    Sum the images in a header according to the idxs_list
+
+    Parameters
+    ----------
+    event_stream: generator
+        The event stream to be summed. The image must be first, with the
+        event itself last
+    idxs_list: list of lists and tuple or list or 'all', optional
+        The list of lists and tuples which specify the images to be summed.
+        If 'all', sum all the images in the run. If None, do nothing.
+        Defaults to None.
+    Yields
+    -------
+    event_stream:
+        The event stream, with the images (in the first position) summed
+
+    >>> from databroker import db
+    >>> hdr = db[-1]
+    >>> total_imgs = sum_images(hdr) # Sum all the images
+    >>> assert len(total_imgs) == 1
+    >>> total_imgs = sum_images(hdr, [1, 2, 3])
+    >>> assert len(total_imgs) == 1
```

Comment:
returns one image that is the sum of the images in image-events (an event that contains an image) 1, 2 and 3 in hdr